### PR TITLE
Add functionality of AI Tutor

### PR DESCRIPTION
**Pull request by hinakhadim**
Add functionality of AI Tutor

### CODE REVIEW

File: ai_tutor/ai_tutor.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    )
+
+    context = String(
+        default="",
+        scope=Scope.settings,
+        help="Write the question context here",
+        multiline=True
+    )
+
+    is_allowed_for_coach_help = Boolean(default=True, scope=Scope.user_state,
```
Comment (ziafazal) [inline ai_tutor/ai_tutor.py]: I think we need two fields one to count number of times learner has asked for review and other fields to set a threshold of how many times review can be asked.

File: ai_tutor/ai_tutor.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    model_name = String(
+        display_name="AI Model Name", values=(
+            'text-davinci-003', 'text-davinci-002', 'text-curie-001', 'text-babbage-001',
+            'text-ada-001'),
+        default="text-davinci-003", scope=Scope.settings,
+        help="Select an AI Text model."
+    )
+
+    description = String(
```
Comment (ziafazal) [inline ai_tutor/ai_tutor.py]: where is this `description` is going to be used?

File: ai_tutor/ai_tutor.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+from xblockutils.studio_editable import StudioEditableXBlockMixin
+import openai
+from django.template import Context, Template
+from django.conf import settings
+
+
+log = logging.getLogger(__name__)
+
+
+class AITutorXBlock(XBlock, StudioEditableXBlockMixin):
```
Comment (ziafazal) [inline ai_tutor/ai_tutor.py]: Inherit it from `CompletableXBlockMixin` and then add
`self.emit_completion(1.0)` inside `submit_answer` method to mark block completion

File: ai_tutor/ai_tutor.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        frag.initialize_js('AITutorXBlock')
+        return frag
+
+    def get_completion(self, prompt='', model='text-davinci-003', temperature=0.5, max_tokens=150, n=1):
+        """
+            Returns the improvement for student answer
+        """
+        try:
+            response = openai.Completion.create(
+                prompt=prompt, model=model, temperature=temperature, max_tokens=max_tokens, n=n)
```
Comment (ziafazal) [inline ai_tutor/ai_tutor.py]: How `max_tokens` and `n` is going to effect the response?

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+setup(
+    name='ai_tutor-xblock',
+    version='0.1',
+    description='ai_tutor XBlock',   # TODO: write a better description.
+    license='UNKNOWN',          # TODO: choose a license: 'AGPL v3' and 'Apache 2.0' are popular.
+    packages=[
+        'ai_tutor',
+    ],
+    install_requires=[
+        'XBlock',
```
Comment (ziafazal) [inline setup.py:43]: `openai` should be in the dependencies of this xblock

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    for root in roots:
+        for dirname, _, files in os.walk(os.path.join(pkg, root)):
+            for fname in files:
+                data.append(os.path.relpath(os.path.join(dirname, fname), pkg))
+
+    return {pkg: data}
+
+
+setup(
+    name='ai_tutor-xblock',
```
Comment (ziafazal) [inline setup.py]: We should be consistent with name of xblock everywhere in the code and files names. what if we use `ai_coach` everywhere?

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    version='0.1',
+    description='ai_tutor XBlock',   # TODO: write a better description.
+    license='UNKNOWN',          # TODO: choose a license: 'AGPL v3' and 'Apache 2.0' are popular.
```
Comment (ziafazal) [inline setup.py]: `AGPL v3` should be fine here.

**Review by ziafazal — CHANGES_REQUESTED** [review]
@hinakhadim thanks for creating this PR. I have some suggestions.

File: ai_tutor/ai_tutor.py (hunk 4 above)
Comment (hinakhadim) [reply ai_tutor/ai_tutor.py]: `max_tokens` is the maximum number of [tokens](https://platform.openai.com/tokenizer) to generate in the completion. Tokens are created from words or common sequences of characters found in text.  `The more the number of token, the more detailed the response is (or response has more words)`

`n` is how many responses will be returned by model.  For example, if ` n= 1` we give prompt='I live in` . It returns [`I live in Lahore`]. but if `n=2`, it returns ['I live in lahore', 'I live in Karachi'] (2 responses). `Warning: The more the value of n, the more it consume your token quota  quickly`.

For detailed information, [OPEN AI Documentation](https://platform.openai.com/docs/api-reference/completions/create)

File: ai_tutor/ai_tutor.py (hunk 2 above)
Comment (hinakhadim) [reply ai_tutor/ai_tutor.py]: The description isn't using anywhere. I created it for if admin wants to store some extra information (regarding question or answer) for himself.

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    for root in roots:
+        for dirname, _, files in os.walk(os.path.join(pkg, root)):
+            for fname in files:
+                data.append(os.path.relpath(os.path.join(dirname, fname), pkg))
+
+    return {pkg: data}
+
+
+setup(
+    name='ai_coach-xblock',
```
Comment (ziafazal) [inline setup.py]: name should be `ai-coach-xblock`

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    version='0.1',
+    description='ai_coach XBlock',
```
Comment (ziafazal) [inline setup.py]: Description should be `AI Coach Xblock evaluates open response answer of a question using Open AI`
and add Long Description should pulled from readme like [this](https://github.com/edly-io/h5pxblock/blob/master/setup.py#L33)

File: README.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
+## AI Tutor XBlock
```
Comment (ziafazal) [inline README.md]: Can we add little description of what this xblock does and possible use cases? Also we should add few screenshots of xblock(in studio and LMS) like [this](https://github.com/edly-io/h5pxblock#readme)

**Review by ziafazal — CHANGES_REQUESTED** [review]
@hinakhadim thanks for addressing feedback its almost there just few minor things to look into.

**Review by ziafazal — APPROVED** [review]
LGTM

**Pull request merged**